File: core/src/main/java/com/timepath/vfs/VFile.java
package com.timepath.vfs;

import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.InputStream;
import java.util.Collection;
import java.util.regex.Pattern;

/**
 * @param <V> Specific type of VFile
 * @author TimePath
 */
public interface VFile<V extends VFile<?>> {

    @NonNls
    String SEPARATOR = "/";

    Pattern SEPARATOR_PATTERN = Pattern.compile(SEPARATOR);

    boolean canExecute();

    boolean canRead();

    boolean canWrite();

    boolean createNewFile();

    boolean delete();

    boolean exists();

    /**
     * @return the file name without {@code SEPARATOR}s
     */
    @NotNull
    String getName();

    @Nullable
    V getParent();

    @NotNull
    Collection<? extends V> list();

    /**
     * Get a file by literal name
     *
     * @param name
     * @return the file, or null
     */
    @Nullable
    V get(@NonNls @NotNull String name);

    @Nullable
    String getPath();

    long getTotalSpace();

    long getUsableSpace();

    boolean isDirectory();

    boolean isFile();

    long lastModified();

    long length();

    boolean renameTo(V dest);

    boolean setExecutable(boolean executable);

    boolean setExecutable(boolean executable, boolean ownerOnly);

    boolean setLastModified(long time);

    boolean setReadable(boolean readable);

    boolean setReadable(boolean readable, boolean ownerOnly);

    boolean setWritable(boolean writable);

    boolean setWritable(boolean writable, boolean ownerOnly);

    /**
     * Open the file for reading. It is the caller's responsibility to close the stream
     *
     * @return the stream, or null
     */
    @Nullable
    InputStream openStream();

    /**
     * @return some identifier, may contain {@code SEPARATOR} unlike {@link #getName()}
     */
    @NotNull
    String toString();
}

File: core/src/main/kotlin/com/timepath/vfs/VFile.kt
package com.timepath.vfs

import org.jetbrains.annotations.NonNls

import java.io.InputStream
import java.util.regex.Pattern

/**
 * @param <V> Specific type of VFile
 * @author TimePath
 */
public trait VFile<V : VFile<V>> {

    fun canExecute(): Boolean

    fun canRead(): Boolean

    fun canWrite(): Boolean

    fun createNewFile(): Boolean

    fun delete(): Boolean

    fun exists(): Boolean

    /**
     * @return the file name without {@code SEPARATOR}s
     */
    val name: String

    val parent: V?

    [suppress("REDUNDANT_PROJECTION")]
    fun list(): Collection<out V>

    /**
     * Get a file by literal name
     *
     * @param name
     * @return the file, or null
     */
    fun get(NonNls name: String): V?

    val path: String

    val totalSpace: Long

    val usableSpace: Long

    val isDirectory: Boolean

    val isFile: Boolean

    val lastModified: Long

    val length: Long

    fun renameTo(dest: V): Boolean

    fun setExecutable(executable: Boolean): Boolean

    fun setExecutable(executable: Boolean, ownerOnly: Boolean): Boolean

    fun setReadable(readable: Boolean): Boolean

    fun setReadable(readable: Boolean, ownerOnly: Boolean): Boolean

    fun setWritable(writable: Boolean): Boolean

    fun setWritable(writable: Boolean, ownerOnly: Boolean): Boolean

    /**
     * Open the file for reading. It is the caller's responsibility to close the stream
     *
     * @return the stream, or null
     */
    fun openStream(): InputStream?

    /**
     * @return some identifier, may contain {@code SEPARATOR} unlike {@link #getName()}
     */
    override fun toString(): String

    class object {

        NonNls
        val SEPARATOR: String = "/"

        val SEPARATOR_PATTERN: Pattern = SEPARATOR.toRegex()
    }
}
